src/Database/LSMTree/Internal/TableHandle.hs

```diff
@@ -0,0 +1,183 @@
+data TableHandle m h = TableHandle {
+    tableSession     :: !(Session m h)
+  , tableConfig      :: !TableConfig
+  , tableWriteBuffer :: !(WriteBuffer)
+    -- | A hierarchy of levels. The vector indexes double as level numbers.
+  , tableLevels      :: !(V.Vector (Level h))
+    -- | Cache of flattened 'levels'.
+    --
+    -- INVARIANT: when 'level's is modified, this cache should be updated as
+    -- well, for example using 'mkLevelsCache'.
+  , tableLevelsCache :: !(LevelsCache h)
+  }
+
+-- | Runs in order from newer to older
+newtype Level h = Level {
+    residentRuns :: V.Vector (Run h)
+{- TODO: this is where ongoing merges appear once we implement scheduling ,
+  incomingRuns :: MergingRun
+-}
+  }
+
+-- | Flattend cache of the runs that referenced by a table handle.
+--
+-- This cache includes a vector of runs, but also vectors of the runs broken
+-- down into components, like bloom filters, fence pointer indexes and file
+-- handles. This allows for quick access in the lookup code. Recomputing this
+-- cache should be relatively rare.
+--
+-- Use 'mkLevelsCache' to ensure that there are no mismatches between the vector
+-- of runs and the vectors of run components.
+data LevelsCache h = LevelsCache_ {
+    cachedRuns      :: !(V.Vector (Run h))
+  , cachedFilters   :: !(V.Vector (Bloom SerialisedKey))
+  , cachedIndexes   :: !(V.Vector IndexCompact)
+  , cachedKOpsFiles :: !(V.Vector h)
+  }
+
+-- | Flatten the argument 'Level's into a single vector of runs, and use that to
+-- populate the 'LevelsCache'.
+mkLevelsCache :: V.Vector (Level h) -> LevelsCache h
+mkLevelsCache lvls = LevelsCache_ {
+      cachedRuns      = rs
+    , cachedFilters   = V.map Run.runFilter rs
+    , cachedIndexes   = V.map Run.runIndex rs
+    , cachedKOpsFiles = V.map Run.runKOpsFile rs
+    }
+  where rs = V.concatMap residentRuns lvls
+
+{-------------------------------------------------------------------------------
+  Configuration
+-------------------------------------------------------------------------------}
+
+-- | Table configuration parameters, including LSM tree tuning parameters.
+--
+-- Some config options are fixed (for now):
+--
+-- * Merge policy: Tiering
+--
+-- * Size ratio: 4
+data TableConfig = TableConfig {
+    confMergePolicy      :: !MergePolicy
+    -- Size ratio between the capacities of adjacent levels.
+  , confSizeRatio        :: !SizeRatio
+    -- | Total number of bytes that the write buffer can use.
+    --
+    -- The maximum is 4GiB, which should be more than enough for realistic
+    -- applications.
+  , confWriteBufferAlloc :: !Word32
+  , confBloomFilterAlloc :: !BloomFilterAlloc
+    -- | Function for resolving 'Mupsert' values. This should be 'Nothing' for
+    -- normal tables.
+  , confResolveMupsert   :: !(Maybe ResolveMupsert)
+  }
+  deriving Show
+
+data MergePolicy =
+    MergePolicyTiering
+{- TODO: disabled for now
+  | MergePolicyLevelling
+-}
```

Comment:
We are not using pure tiering. I think the name for having levelling on the last level was "lazy levelling", so should we use that here?